Foundation/URLSession/NativeProtocol.swift

```diff
@@ -0,0 +1,754 @@
+        self.requestBodySource = bodySource
+        self.bodyDataDrain = bodyDataDrain
+    }
+    
+}
+
+struct _Delimiters {
+    /// *Carriage Return* symbol
+    static let CR: UInt8 = 0x0d
+    /// *Line Feed* symbol
+    static let LF: UInt8 = 0x0a
+    /// *Space* symbol
+    static let Space = UnicodeScalar(0x20)
+    static let HorizontalTab = UnicodeScalar(0x09)
+    static let Colon = UnicodeScalar(0x3a)
+    /// *Separators* according to RFC 2616
+    static let Separators = NSCharacterSet(charactersIn: "()<>@,;:\\\"/[]?={} \t")
+}
+
+extension _NativeProtocol {
+    /// An HTTP header being parsed.
+    ///
+    /// It can either be complete (i.e. the final CR LF CR LF has been
+    /// received), or partial.
+    internal enum _ParsedResponseHeader {
+        case partial(_ResponseHeaderLines)
+        case complete(_ResponseHeaderLines)
+        init() {
+            self = .partial(_ResponseHeaderLines())
+        }
+    }
+    /// A type safe wrapper around multiple lines of headers.
+    ///
+    /// This can be converted into an `NSHTTPURLResponse`.
+    internal struct _ResponseHeaderLines {
+        let lines: [String]
+        init() {
+            self.lines = []
+        }
+        init(headerLines: [String]) {
+            self.lines = headerLines
+        }
+    }
+    
+}
+
+extension _NativeProtocol._ResponseHeaderLines {
+    
+    func createURLResponse(for URL: URL, contentLength: Int64) -> URLResponse? {
+        return URLResponse(url: URL, mimeType: nil, expectedContentLength: Int(contentLength),textEncodingName: nil)
+    }
+}
+
+extension _NativeProtocol {
+        /// Set request body length.
+        ///
+        /// An unknown length
+        func set(requestBodyLength length: _HTTPURLProtocol._RequestBodyLength) {
+            switch length {
+            case .noBody:
+                easyHandle.set(upload: false)
+                easyHandle.set(requestBodyLength: 0)
+            case .length(let length):
+                easyHandle.set(upload: true)
+                easyHandle.set(requestBodyLength: Int64(length))
+            case .unknown:
+                easyHandle.set(upload: true)
+                easyHandle.set(requestBodyLength: -1)
+            }
+        }
+        enum _RequestBodyLength {
+            case noBody
+            ///
+            case length(UInt64)
+            /// Will result in a chunked upload
+            case unknown
+        }
+    }
+
+extension _NativeProtocol._ParsedResponseHeader {
+    
+    /// Parse a header line passed by libcurl.
+    ///
+    /// These contain the <CRLF> ending and the final line contains nothing but
+    /// that ending.
+    /// - Returns: Returning nil indicates failure. Otherwise returns a new
+    ///     `ParsedResponseHeader` with the given line added.
+    func byAppending(headerLine data: Data, headerCompleted: (String) -> Bool) -> _NativeProtocol._ParsedResponseHeader? {
+        // The buffer must end in CRLF
```

Comment:
What happens to the data if there isn't a CRLF at the end, such as when you've read a partial headerLine, or in case it contains multiple headerLines?